[V2V] Add CPU and network throttling in model

fdupont-redhat opened the pull request:
Now that we introduced default values for CPU and network limits in the settings and that the throttling mechanism is backported into the model, it is time to add the model code to leverage the CPU and network limits at the conversion host level.

This pull request adds:

- New methods in the ConversionHost class to compute the CPU and network limits. It divides the limit per host by the number of active migration tasks.
- New method in the ConversionHost class to apply the limits for a specific task. It creates a throttling file on the conversion host. The path of the file is provided by virt-v2v-wrapper and stored in the task.
- New method in the ServiceTemplateTransformationPlanTask to apply the limits. It collects all the limits in a hash and asks the conversion host to write them in the throttling file.
- Extend the InfraConversionJob class to ask the task to apply the limits. This allows to revisit the limits on every poll.

Currently all limits are at the conversion host level, so it looks like the apply_virtv2v_limits method should belong to the ConversionHost class, but later we may have limits on different scopes. This allows better flexibility to concentrate the hash build in the task.

Associated RHBZ: https://bugzilla.redhat.com/show_bug.cgi?id=1690851
Depends On: https://github.com/ManageIQ/manageiq-gems-pending/pull/426

fdupont-redhat commented:
@miq-bot add_label transformation, enhancement, hammer/yes, wip
@miq-bot add_reviewer @jameswnl 
@miq-bot add_reviewer @djberg96

djberg96 commented on app/models/conversion_host.rb:
```diff
@@ -56,6 +56,22 @@ def ipaddress(family = 'ipv4')
+    value == 'unlimited' ? value : "#{value.to_i / active_tasks.size}"
+  end
+
+  def apply_virtv2v_limits(path, limits)
```
Are there any reasonable defaults we can set for `path` and/or `limits`?

fdupont-redhat replied:
Nope. Path is stored in the task and specific to a single migration, so no default. And limits is built by the task object from different sources, so the default can be `{}`.

djberg96 commented on app/models/conversion_host.rb:
```diff
@@ -56,6 +56,22 @@ def ipaddress(family = 'ipv4')
     resource.ipaddresses.detect { |ip| IPAddr.new(ip).send("#{family}?") }
   end
 
+  def get_cpu_limit
```
Do we need `get_`, or can we just redefine the existing method?

jameswnl replied:
`Setting(s)`?

fdupont-redhat replied:
I'm neither a big fan of `get_`. It's an attribute of ConversionHost, but should default to `Settings.transformation.limits.cpu_limit_per_host` which we implemented in a previous PR. Can we override an attribute with a method ?

djberg96 replied:
Yep, I think this is the way: http://joshfrankel.me/blog/overriding-activerecord-getters-with-the-same-attribute-name/

djberg96 commented on app/models/infra_conversion_job.rb:
```diff
@@ -104,6 +104,11 @@ def poll_conversion
     case v2v_status
     when 'active'
+      begin
+        migration_task.apply_virtv2v_limits if migration_task.options.fetch_path(:virtv2v_wrapper, 'throttling_file')
```
Where is `throttling_file` coming from? Is that a configurable thing? Or is hard coding it ok?

fdupont-redhat replied:
It comes from the output of virt-v2v-wrapper, like `state_file`. So, pretty much hard coded.

jameswnl replied:
@fdupont-redhat, 2 question
* are we doing this `apply_virtv2v_limits` in every polling, like every 15s, regardless limits changed or not?
* so this control flow is `infra_conversion_job -> miq_request_task.apply_virtv2v_limits -> conversion_host.apply_virtv2v_limits`? seems convoluted. This is the reason I asked to re-consider where we want to keep the throttling logic.

djberg96 replied:
Agree with @jameswnl Seems like too many hoops.

jameswnl commented on app/models/conversion_host.rb:
```diff
@@ -56,6 +56,22 @@ def ipaddress(family = 'ipv4')
+  end
+
+  def get_network_limit
+    value = network_limit || Setting.transformation.limits.network_limit_per_host
```
Setting`s`

jameswnl commented on app/models/service_template_transformation_plan_task.rb:
```diff
@@ -233,6 +233,16 @@ def get_conversion_state
 
+  # Applies the limits for the task.
+  # CPU and network limits are set at the host level but other limits may not.
+  def apply_virtv2v_limits
```
I am thinking where this and forthcoming throttling code should reside.
May be a `mixin` or a sub-module of `infra_conversion_job`?

fdupont-redhat replied:
For me, the logic should reside in the task as it has all the knowledge of the execution context. The infra_conversion_job is mainly a state machine, and should remain pretty _stupid_ and delegate any _intelligence_ to the task. As whether it should be in a mixin, I'm wondering how much code can be reused by other classes.

jameswnl commented on app/models/service_template_transformation_plan_task.rb:
```diff
@@ -233,6 +233,16 @@ def get_conversion_state
+      :cpu     => conversion_host.get_cpu_limit,
+      :network => conversion_host.get_network_limit
+    }
+    conversion_host.apply_virtv2v_limits(options[:virtv2v_wrapper]['throttling_file'], limits)
```
This construction of `limits` seems redundant. If this `self` object has no opinion on how the limits are constructed, why not let `conversion_host` do it internally?

djberg96 replied:
I think I kind of agree with @jameswnl. Seems like this could be refactored away or delegated.

agrare commented:
hey guys what is the status of this?  I don't see any updates for a couple of weeks and it is still WIP

mturley commented:
@fdupont-redhat @djberg96 @jameswnl any updates on this one? We're holding off on merging https://github.com/ManageIQ/manageiq-v2v/pull/915 until it's ready

fdupont-redhat commented:
@djberg96 @jameswnl I took time to revisit and rethink this, and I moved the limits calculation into InfraConversionThrottler class, and added a dispatch to update and apply the limits to all jobs. This simplifies the design. I still need to add the specs.

fdupont-redhat commented:
@miq-bot remove-label wip
@djberg96 @agrare could you review, please ?

djberg96 commented on lib/infra_conversion_throttler.rb:
```diff
@@ -20,25 +20,41 @@ def self.pending_conversion_jobs
     pending.group_by { |job| job.migration_task.destination_ems }
   end
 
+  # @return [Hash] the list of jobs in state 'running', grouped by conversion host
```
Could the state also be 'migrate' here?

fdupont-redhat replied:
Nope. `migrate` state is on the ServiceTemplateTransformationPlanTask, not the InfraConversionJob.

fdupont-redhat commented:
@miq-bot add-reviewer @agrare

agrare commented on lib/infra_conversion_throttler.rb:
```diff
@@ -19,4 +19,43 @@ def self.pending_conversion_jobs
+  # Applying the limits is done via the conversion_host which handles the writing.
+  def self.apply_limits
+    running_conversion_jobs.each do |ch, jobs|
+      number_of_jobs = ch.active_tasks.size
```
Hm if ConversionHost#active_tasks got out of sync with the running jobs somehow and was 0 then [this division](https://github.com/ManageIQ/manageiq/pull/18576/files#diff-8dffd85ef5e1772dce26180aa8ff6d65R51) will blow up...maybe set this to 1 if it is 0?

fdupont-redhat replied:
Good catch. `jobs.size` is safer because `jobs` can't be empty inside the loop.

agrare commented on lib/infra_conversion_throttler.rb:
```diff
@@ -19,4 +19,43 @@ def self.pending_conversion_jobs
+        throttling_file_path = migration_task.options.fetch_path(:virtv2v_wrapper, 'throttling_file')
+        next unless throttling_file_path
+        limits = {
+          :cpu     => cpu_limit == 'unlimited' ? cpu_limit : (cpu_limit.to_i / number_of_jobs).to_s,
```
It is minor but this calculation doesn't depend on anything in this loop so you could calculate it [up here](https://github.com/ManageIQ/manageiq/pull/18576/files#diff-8dffd85ef5e1772dce26180aa8ff6d65R44)

Something like
```
cpu_limit = ch.cpu_limit || Settings.transformation.limits.cpu_limit_per_host
cpu_limit = (cpu_limit.to_i / number_of_jobs).to_s unless cpu_limit == "unlimited"
```

fdupont-redhat replied:
Moved.

agrare commented on lib/infra_conversion_throttler.rb:
```diff
@@ -40,16 +40,21 @@ def self.running_conversion_jobs
   def self.apply_limits
     running_conversion_jobs.each do |ch, jobs|
-      number_of_jobs = ch.active_tasks.size
+      number_of_jobs = jobs.size
```
:+1: even better

simaishi commented:
Hammer backport details:

```
$ git log -1
commit b34ee8b44080745a2441cbc3bf3be890258637d3
Author: Adam Grare <agrare@redhat.com>
Date:   Thu Apr 18 14:15:24 2019 -0400

    Merge pull request #18576 from fdupont-redhat/v2v_apply_resource_limits_to_virtv2v
    
    [V2V] Add CPU and network throttling in model
    
    (cherry picked from commit ce7c67c404127843801f22471f1d286383a9ef17)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1702085
```

simaishi commented:
Reverted Hammer backport

```
commit 9d0322d76c646d4f3a15fc36b10caa6972b013b7
Author: Satoe Imaishi <simaishi@redhat.com>
Date:   Tue May 7 18:26:33 2019 -0400

    Revert "Merge pull request #18576 from fdupont-redhat/v2v_apply_resource_limits_to_virtv2v"
    
    This reverts commit b34ee8b44080745a2441cbc3bf3be890258637d3.
```
